### ISSUE TITLE
Support aarch64 sendmmsg, recvmmsg and YIELD

### ISSUE THREAD

**Issue by ChristopherWirt** [netty/netty#11441]
### Expected behaviour
Equivalent network performance on aarch64

### Actual behaviour
Slower than I was hoping when under load.

We deploy a high network throughput application using Netty w/Epoll on x86. I’ve been testing out arm/aarch64 as a possible replacement as we’ve seen great results with some of our compute intensive applications. Sadly, I’ve been having some performance issues in doing so and slowly been digging into why this might be. About 80% of our CPU time is in Netty land regardless of platform, and workloads are identical.

We're see higher rescheduling interrupts (roughly 8x) and much higher CPU time spent in `__softirqentry_text_start`, `__wake_up_common_lock` and `arch_local_irq_enable` on aarch64 than x86. We also see overall higher p95,p99 latencies once we start getting up to high CPU usage.

This has led me to https://github.com/netty/netty/blob/4.1/transport-native-epoll/src/main/c/netty_epoll_native.c

I noticed that netty_epoll_native.c defines SYS_recvmmsg and SYS_sendmmsg for x86, but not aarch64
https://github.com/netty/netty/blob/4.1/transport-native-epoll/src/main/c/netty_epoll_native.c#L98

It also adds an asm pause - when compiled for x86 in cpu_relax()
https://github.com/netty/netty/blob/4.1/transport-native-epoll/src/main/c/netty_epoll_native.c#L295

I’m barely c and asm literate, but at this surface this seems just like a case of adding a few blocks of code

```
#elif defined(__aarch64__)
// See https://thog.github.io/syscalls-table-aarch64/latest.html
#define SYS_recvmmsg 365
....
#elif defined(__aarch64__)
// See https://thog.github.io/syscalls-table-aarch64/latest.html
#define SYS_sendmmsg 296
```

And maybe 
```
#if defined(__aarch64__)
    asm volatile("yield\n": : :"memory");
#endif
```

I haven't tested this yet or even written anything that compiles. Literally only checked out the Netty code a few hours ago.

What do you think the odds of this helping/solving my issues are? Is anyone already looking at doing this? or experienced similar issues?

Thanks

### Netty version
4.1.65.Final

### JVM version (e.g. `java -version`)
openjdk version "11.0.11" 2021-04-20 LTS
OpenJDK Runtime Environment Corretto-11.0.11.9.1 (build 11.0.11+9-LTS)
OpenJDK 64-Bit Server VM Corretto-11.0.11.9.1 (build 11.0.11+9-LTS, mixed mode)

### OS version (e.g. `uname -a`)
Linux 5.10.35-31.135.amzn2.aarch64 #1 SMP Tue May 11 22:05:38 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux

**Comment by normanmaurer** [comment]
@ChristopherWirt this is a good catch... You want to open a PR to add:

```
#elif defined(__aarch64__)
// See https://thog.github.io/syscalls-table-aarch64/latest.html
#define SYS_recvmmsg 365
....
#elif defined(__aarch64__)
// See https://thog.github.io/syscalls-table-aarch64/latest.html
#define SYS_sendmmsg 296
```
 ?

I would be happy to review it.

I would do this first and then look into the volatile as a followup

**Comment by ChristopherWirt** [comment]
Will do. Thanks

**Comment by AGSaidi** [comment]
With the `SYS_sendmsg` i don't believe the code would work at all if they weren't defined. I'm guessing these existed on x86 to enable building on platforms with very old libc/headers versions that didn't define these constants. Given the age of the system calls and that they've probably been around before 64b Arm was defined these defined aren't needed.

The change in spin behavior might improve performance here, however with a `yield` it won't help. On most Arm processors a `yield` is a `nop` while `pause` tends to cause a 30-100 cycle delay. We've found `isb` on Arm ends up producing delays that are closer to `pause` x86 in other projects (e.g. MySQL), so there is a chance this will improve the rescheduling behavior you see. 

There are also similar changes that have just went into Corretto [here](https://github.com/corretto/corretto-11/commit/dfb4965877a5810011514bd9294175eccd4b6d0d) and [here](https://github.com/corretto/corretto-11/commit/a49a79bb2e7ac4a2265c51c9fb1c3fcf90dc7c9d) that will be included in the next version of Corretto released in July.

**Comment by normanmaurer** [comment]
Actually @AGSaidi is right ... @ChristopherWirt so this can't be the reason really

**Comment by AGSaidi** [comment]
@ChristopherWirt is there a small reproducer you see the problem with you can share?

**Comment by ChristopherWirt** [comment]
> With the SYS_sendmsg i don't believe the code would work at all if they weren't defined.

@AGSaidi Do you mean that without `SYS_sendmsg` defined for aarch64, it's fine because it does the right thing anyway? we just spend a little longer in Javacode and call a different Socket method? So unlikely the cause of my troubles. Which I think makes sense.

RE: the explicit pause in cpu_relax() - It seems like it's still worth trying 'isb' still?

I don't have a reproducer at the moment.

**Comment by normanmaurer** [comment]
> > With the SYS_sendmsg i don't believe the code would work at all if they weren't defined.
> 
> @AGSaidi Do you mean that without `SYS_sendmsg` explicitly defined for aarch64, it's fine because it does the right thing anyway? we just spend a little longer in Javacode and call a different Socket method? So unlikely the cause of my troubles. Which I think makes sense.

Yes even without this definition it should do the correct thing. 

> 
> RE: the explicit pause in cpu_relax() - It seems like it's still worth trying 'isb' still?
> 
> I don't have a reproducer at the moment.

**Comment by AGSaidi** [comment]
> RE: the explicit pause in cpu_relax() - It seems like it's still worth trying 'isb' still?

Yes, it might make a difference.

**Comment by normanmaurer** [comment]
That said this is only for "fast poll" which is not used by default so I think that is not what affect you.

**Comment by normanmaurer** [comment]
Any updates here ?
